documentation/p0201.md

```diff
@@ -1,828 +1,828 @@
+  creates an object that owns the object `*p`,  with a copier of
+  type C initialized from `std::move(c)` and a deleter of type D
+  initialized from `std::move(d)`.
+
+* Throws: `bad_alloc` if required storage cannot be obtained;
+  `bad_polymorphic_value_construction` if `is_same_v<C,
+  default_copy<U>>`, `is_same_v<D, default_delete<U>>` and
+  `typeid(*p)!=typeid(U)` are all `true`.
+
+* Postconditions: If `p` is null, the empty object created has no copier and no deleter.
+  Otherwise the object created owns the object `*p` and has a copier and a deleter 
+  present, initialized from `std::move(c)` and `std::move(d)` respectively.
+
+```cpp
+polymorphic_value(const polymorphic_value& pv);
+template <class U> explicit polymorphic_value(const polymorphic_value<U>& pv);
+```
+
+* Constraints: For the second constructor, `is_convertible_v<U*, T*>` is true.
+
+* Effects: If `pv` is empty, constructs an empty object. Otherwise
+  creates an object that owns a copy of the object managed by `pv`. 
+  If a copier and deleter are present in `pv` then 
+  the copy is created by the copier in `pv`. Otherwise the copy is created by copy 
+  construction of the owned object.  If a copier and deleter are present 
+  in `pv` then the copier and deleter of the object constructed 
+  are copied from those in `pv`.
+
+* Postconditions:  `bool(*this) == bool(pv)`.
+
+* Throws: Any exception thrown by invocation of the copier, copying the 
+  copier and deleter, or `bad_alloc` if required storage cannot be obtained.
+
+```cpp
+polymorphic_value(polymorphic_value&& pv) noexcept;
+template <class U> explicit polymorphic_value(polymorphic_value<U>&& pv);
+```
+
+* Constraints: For the second constructor, `is_convertible_v<U*, T*>` is true.
+
+* Effects: If `pv` is empty, constructs an empty object. Otherwise the object
+  owned by `pv` is transferred to the constructed object. If a copier and deleter
+  are present in `pv` then the copier and deleter are transferred to the 
+  constructed object.
+
+* Postconditions:  `*this` owns the object previously owned by `pv` (if any). 
+  `pv` is empty.
+
+### X.Z.4 Class template `polymorphic_value` destructor [polymorphic_value.dtor]
+
+```cpp
+~polymorphic_value();
+```
+
+* Effects: If a copier `c` and a deleter `d` are present, 
+evaluates `d(operator->())` and destroys `c` and `d`. 
+Otherwise destroys the owned object (if any) and frees its storage.
+
+### X.Z.5 Class template `polymorphic_value` assignment [polymorphic_value.assignment]
+
+```cpp
+polymorphic_value& operator=(const polymorphic_value& pv);
+```
+
+* Effects: Equivalent to `polymorphic_value(pv).swap(*this)`. 
+  No effects if an exception is thrown. 
+
+* Throws: Any exception thrown by the copier or `bad_alloc` if required
+  storage cannot be obtained.
+
+* Returns: `*this`.
+
+* Postconditions:  The state of `*this` is as if copy constructed from `pv`.
+
+```cpp
+polymorphic_value& operator=(polymorphic_value&& pv) noexcept;

```

Comment:
@jbcoe Please can you double check this, we have multiple comments from both J.Wakely and D.Krügler indicating the is a problem here but I cannot see it:

- J.Wakely
> move assignment operator should use std::move(pv).

- D.Krügler
> The function is specified to be noexcept but has the following Effects: Equivalent to polymorphic_value(pv).swap(*this). These Effects say that it would call the copy constructor (which may throw). I assume that the intention is to call the move constructor, but that would require to reword the Effects to say instead: Equivalent to polymorphic_value(std::move(pv)).swap(*this).

Reply:
Effects as written look good to me.